### ISSUE TITLE
Segmentation Fault during repeated execution of circuit using `density_matric_gpu` simulator)

### ISSUE THREAD

**Issue by we-taper** [Qiskit/qiskit-aer#755]
### Informations

- **Qiskit Aer version**:
```python
{'qiskit-terra': '0.14.1',
 'qiskit-aer': '0.5.1',
 'qiskit-ignis': '0.3.0',
 'qiskit-ibmq-provider': '0.7.1',
 'qiskit-aqua': '0.7.1',
 'qiskit': '0.19.2'}
```
- **Python version**: 3.6.9 
- **Operating system**: Ubuntu 19.04

### What is the current behavior?

After repeated execution of circuits (a common scenario found in variational algorithm),
- the density matrix CPU simulator will cause `segmentation fault (core dumped)`
- the density matric GPU simulator will cause 
```bash
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  parallel_for failed: cudaErrorInvalidConfiguration: invalid configuration argument
[1]    11839 abort (core dumped)
```

### Steps to reproduce the problem

A example script:

```python
import qiskit as qk
import time

qc = qk.QuantumCircuit(20, 1)
[qc.h(_) for _ in range(20)]
qc.measure(0, 0)
bkd = qk.Aer.get_backend("qasm_simulator")
option = {"method": "density_matrix_gpu"}  # or density_matrix_cpu
i = 0
while True:
    i += 1
    print("Looping: ", i)
    qk.execute(qc, bkd, backend_options=option)
```
The error typically happen around 100 to 200 loops.

### What is the expected behavior?

The simulator should be able to indefinitely execute the script.

### Suggested solutions

Not sure.

**Comment by chriseclectic** [comment]
Your test script is basically launching and backgrounding infinite simulation job subprocesses through python multiprocessing. And for this particular example of a 20 qubit circuit on density matrix simulator each of these concurrently running jobs would require ~18 TB of memory to store the state.

So I think the true issue here is each process is not being terminated with en error for insufficient memory.

As an aside for local simulation you should make sure you block on the job result in the while look so you aren't spinning up an infinite number of async simulation jobs through python. Also parallel execution should be handled through parallel circuit execution of the qobj, otherwise thread resources will not be managed correctly.

**Comment by we-taper** [comment]
Hi @chriseclectic, thanks for the reply. I understand that 20 qubits is not a good example (and I should have called `.result()` to enforce that the script got run. 

But if we have a 10 qubit example, the same loop failed after around 200 loops with the error message: 
```python
QiskitError: 'PARTIAL COMPLETED ,  ERROR: std::bad_alloc: cudaErrorMemoryAllocation: out of memory'
```

**Comment by chriseclectic** [comment]
There was a memory leak in the GPU simulator that has been fixed in master (#717 ), that might be the cause for the cuda error you were seeing? This has been fixed in master, but is not in the stable release yet. We are working on a patch release at the moment that might include some of these things.

I will also  try and make a fix for the issue of not raising an error if there is insufficient memory for the CPU case.

**Comment by we-taper** [comment]
Great! Looking forward to the patch!
